### PR TITLE
Add DatasetInfo and refactor dataset base classes 

### PR DESCRIPTION
**Pull request by ly015**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This PR is to merge the changes of #663 into master.

## Modification

Please refer to the original PR for details.

### 1 DatasetInfo is needed to build dataset and inference model

`DatasetInfo` is defined in `mmpose/datasets/dataset_info.py` and is used to parse and hold dataset meta information (e.g. keypoint definition, skeleton). It is usually initialized from a config in `configs/_base_/datasets/`.

Please note that the old-fashioned usage listed below is still supported with warning information.

#### 1.1 Build dataset

```python
# New style
dataset_info = Dataset(dataset_info_cfg)
dataset = TopDownCOCODataset(
        ann_file='tests/data/coco/test_coco.json',
        img_prefix='tests/data/coco/',
        data_cfg=data_cfg_copy,
        pipeline=[],
        dataset_info=dataset_info,
        test_mode=True)

# Old style
dataset = TopDownCOCODataset(
        ann_file='tests/data/coco/test_coco.json',
        img_prefix='tests/data/coco/',
        data_cfg=data_cfg_copy,
        pipeline=[],
        test_mode=True)
```

Users won't have to worry about the difference in most cases, because it is encapsulated by high-level APIs like `mmpose.apis.build_dataset`.

#### 1.2 Inference model

```python
## New style
# test a single image, with a list of bboxes.
pose_results, _ = inference_top_down_pose_model(
        pose_model,
        image_name,
        person_result,
        format='xywh',
        dataset_info=dataset_info)
# show the results
vis_pose_result(
    pose_model, image_name, pose_results, dataset_info=dataset_info)

## Old style
# test a single image, with a list of bboxes.
pose_results, _ = inference_top_down_pose_model(
        pose_model,
        image_name,
        person_result,
        format='xywh',
        dataset='TopDownCOCODataset')
# show the results
vis_pose_result(
    pose_model, image_name, pose_results, dataset='TopDownCOCODataset') 
```

### 2 Use new base classes of datasets

Base classes for datasets are placed at `mmpose/datasets/datasets/base`. Please check the following comparison table for new and old class names.

| New Base Class | Old Base Class|
| -- | -- |
|Kpt2dSviewRgbImgTopDownDataset|TopDownBaseDataset, HandBaseDataset, AnimalBaseDataset, FaceBaseDataset, FashionBaseDataset|
|Kpt2dSviewRgbImgBottomUpDataset|BottomUpBaseDataset|
|Kpt3dSviewKpt2dDataset|Body3DBaseDataset|
|Kpt3dSviewRgbImgTopDownDataset|HandBaseDataset\*|

\* `InterHand3DDataset`, which was originally based on `HandBaseDataset`, is based `Kpt3dSviewRgbImgTopDownDataset` in this PR. Other hand datasets are based on `Kpt2dSviewRgbImgTopDownDataset`.

### 3 Use `pose_link_color` to replace `pose_limb_color`

```python
# New style
img_vis_2d = imshow_keypoints(
    img,
    [kpts],
    skeleton=skeleton,
    pose_kpt_color=[(127, 127, 127)* len(kpts)],
    pose_link_color=[(127, 127, 127)* len(skeletons)])

# New style
img_vis_2d = imshow_keypoints(
    img,
    [kpts],
    skeleton=skeleton,
    pose_kpt_color=[(127, 127, 127)* len(kpts)],
    pose_limb_color=[(127, 127, 127)* len(skeletons)])
```

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: .dev_scripts/benchmark/benchmark_regression.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                          f'CPUS_PER_TASK={cur_cpus_per_task} ' + \
+                          f'./tools/slurm_train.sh {cur_partition} ' + \
+                          f'{cur_task_name} ' + \
+                          f'{cur_config} {cur_work_dir} ' + \
+                          f'{py_cmd}'
+                    os.system(f'tmux send-keys "{cmd}" "C-m"')
+
+                cur_port += 1
+                # if the port is used, use a random number for port
+                while not is_port_available(cur_port, ip):
```
Comment (innerlee) [inline .dev_scripts/benchmark/benchmark_regression.py:236]: How about wrap L236-237 into a util function `get_available_port(range=(29000, 39000))`?
Maybe we can also improve the train.py and test.py?

File: .dev_scripts/github/update_copyright.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python
```
Comment (innerlee) [inline .dev_scripts/github/update_copyright.py:1]: These scripts do not belong to this pr. Maybe do a proper rebase?

Comment (ly015) [reply .dev_scripts/github/update_copyright.py:1]: Yes, please wait a moment and I will rebase it to the latest master.

**Review by jin-s13 — APPROVED** [review]

**Comment by innerlee** [comment]
We may add a section in the doc, describing how we support datasets via config, and how to define customized datasets
